### PR TITLE
Implement disjunctive normal form (DNF) transform.

### PR DESCRIPTION
**Pull request by ehsannas**

### CODE REVIEW

**Review by schmidt-sebastian — COMMENTED** [review]
FYI: I will review this PR for style nits and whitespacing issues, but I think I need to a document to explain how this is meant to work before I can provide more meaningful feedback.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/core/FilterTest.java [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+  public void testComputeDnf4() {
+    // A | (B & C) | ( ((D)) | (E | F) | (G & H) ) --> A | (B & C) | D | E | F | (G & H)
+    CompositeFilter compositeFilter =
+        orFilters(
+            A,
+            andFilters(B, C),
+            orFilters(andFilters(orFilters(D)), orFilters(E, F), andFilters(G, H)));
+    Filter[] expectedDnfTerms = {A, andFilters(B, C), D, E, F, andFilters(G, H)};
+    CompositeFilter expectedResult = orFilters(expectedDnfTerms);
+    assertThat(computeDnf(compositeFilter)).isEqualTo(expectedResult);
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/core/FilterTest.java:242]: In general, we should only be testing public APIs. This makes it harder to test the logic in complicated helper methods, so it the test cases above it makes sense that you directly invoke `applyX()`. I wonder though if it makes sense in the tests here as you are invoking both the public and non-public API.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/core/FilterTest.java [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+
+  @Test
+  public void testComputeDnf5() {
+    //    A & (B | C) & ( ((D)) & (E | F) & (G & H) )
+    // -> A & (B | C) & D & (E | F) & G & H
+    // -> ((A & B) | (A & C)) & D & (E | F) & G & H
+    // -> ((A & B & D) | (A & C & D)) & (E|F) & G & H
+    // -> ((A & B & D & E) | (A & B & D & F) | (A & C & D & E) | (A & C & D & F)) & G & H
+    // -> ((A&B&D&E&G) | (A & B & D & F & G) | (A & C & D & E & G) | (A & C & D & F & G)) & H
+    // -> (A&B&D&E&G&H) | (A&B&D&F&G&H) | (A & C & D & E & G & H) | (A & C & D & F & G & H)
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/core/FilterTest.java:254]: Fun :)

File: firebase-firestore/src/test/java/com/google/firebase/firestore/core/FilterTest.java [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+    CompositeFilter compositeFilter =
+        andFilters(
+            A,
+            orFilters(B, C),
+            andFilters(andFilters(orFilters(D)), orFilters(E, F), andFilters(G, H)));
+    Filter[] expectedDnfTerms = {
+      andFilters(D, E, G, H, A, B),
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/core/FilterTest.java:261]: Re-order to match comment?

File: firebase-firestore/src/test/java/com/google/firebase/firestore/core/QueryTest.java [hunk 4]
```diff
@@ -689,4 +691,76 @@ public void testCanonicalIdsAreStable() {
...
+  public void testOrQuery() {
+    MutableDocument doc1 = doc("collection/1", 0, map("a", 1, "b", 0));
+    MutableDocument doc2 = doc("collection/2", 0, map("a", 2, "b", 1));
+    MutableDocument doc3 = doc("collection/3", 0, map("a", 3, "b", 2));
+    MutableDocument doc4 = doc("collection/4", 0, map("a", 1, "b", 3));
+    MutableDocument doc5 = doc("collection/5", 0, map("a", 1, "b", 1));
+
+    // Two equalities: a==1 || b==1.
+    Query query1 =
+        Query.atPath(ResourcePath.fromString("collection"))
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/core/QueryTest.java:705]: `query('collection)` (there is a helper)

File: firebase-firestore/src/test/java/com/google/firebase/firestore/core/QueryTest.java [hunk 5]
```diff
@@ -689,4 +691,76 @@ public void testCanonicalIdsAreStable() {
...
+            .filter(orFilters(filter("a", "==", 1), filter("b", "==", 1)));
+    assertQueryMatches(
+        query1,
+        /* match */ Arrays.asList(doc1, doc2, doc4, doc5),
+        /* not match */ Arrays.asList(doc3));
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/core/QueryTest.java:710]: Unrelated thought: What should the canonical ID of a DNF-query and an equivalent non-DNF query be? Should they match? Should we create one or two listens?

File: firebase-firestore/src/test/java/com/google/firebase/firestore/core/QueryTest.java [hunk 6]
```diff
@@ -689,4 +691,76 @@ public void testCanonicalIdsAreStable() {
...
+                    orFilters(filter("a", "==", 2), filter("b", "==", 2)),
+                    orFilters(filter("a", "==", 3), filter("b", "==", 3))));
+    assertQueryMatches(
+        query5,
+        /* match */ Arrays.asList(doc3),
+        /* not match */ Arrays.asList(doc1, doc2, doc4, doc5));
+  }
+
+  private void assertQueryMatches(
+      Query query, List<MutableDocument> match, List<MutableDocument> nonmatch) {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/core/QueryTest.java:758]:
```suggestion
      Query query, List<MutableDocument> matching, List<MutableDocument> nonMatching) {
```

File: firebase-firestore/src/test/java/com/google/firebase/firestore/local/QueryEngineTestCase.java [hunk 7]
```diff
@@ -427,4 +429,58 @@ public void doesNotIncludeDocumentsDeletedByMutation() throws Exception {
...
+    // (a==2 || b==2) && (a==3 || b==3)
+    Query query5 =
+        query("coll")
+            .filter(
+                andFilters(
+                    orFilters(filter("a", "==", 2), filter("b", "==", 2)),
+                    orFilters(filter("a", "==", 3), filter("b", "==", 3))));
+    DocumentSet result5 =
+        expectFullCollectionScan(() -> runQuery(query5, MISSING_LAST_LIMBO_FREE_SNAPSHOT));
+    assertEquals(docSet(query5.comparator(), doc3), result5);
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/local/QueryEngineTestCase.java:484]: This test just tests the same code as the test in the file above, doesn't it? Neither code actually uses the DNF transform. Should we remove this test?

**Review by schmidt-sebastian — COMMENTED** [review]
This is great. Maybe I don't need a design doc after all - but others might still profit from it.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/core/QueryTest.java (hunk 5 above)
Comment (ehsannas) [reply firebase-firestore/src/test/java/com/google/firebase/firestore/core/QueryTest.java:710]: For the time being, the canonical ID of `a || b` is not the same as `b || a`. We can discuss this further, and address it in the next PR.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java [hunk 8]
```diff
@@ -135,10 +148,12 @@ public String getCanonicalId() {
...
 
     List<String> canonicalIds = new ArrayList<>();
-    for (Filter filter : filters) canonicalIds.add(filter.getCanonicalId());
+    for (Filter filter : filters) {
+      canonicalIds.add(filter.getCanonicalId());
+    }
     StringBuilder builder = new StringBuilder();
     builder.append(isConjunction() ? "and(" : "or(");
-    TextUtils.join(",", canonicalIds);
+    builder.append(TextUtils.join(",", canonicalIds));
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java:156]: I'll defer this until we make a decision about whether we want to make post-dnf canonical IDs for composite filters.

[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java:156]: We can defer this, but this is slightly unrelated. toString() is merely for debugging. We could still change the actual canonical ID even if we used it in toString().

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteIndexManager.java [hunk 9]
```diff
@@ -324,7 +324,7 @@ public IndexOffset getMinOffset(Target target) {
     } else {
       // There is an implicit AND operation between all the filters stored in the target.
       List<Filter> dnf =
-          LogicUtils.DnfTransform(
+          LogicUtils.getDnfTerms(
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteIndexManager.java:327]: This should be `getDnfTerms` per the style guide.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/util/LogicUtils.java [hunk 10]
```diff
@@ -24,6 +32,278 @@
...
+  /** Asserts that the given filter is a FieldFilter or CompositeFilter. */
+  private static void assertFieldFilterOrCompositeFilter(Filter filter) {
+    hardAssert(
+        filter instanceof FieldFilter || filter instanceof CompositeFilter,
+        "Only field filters and composite filters are accepted.");
+  }
+
+  /** Returns true if the given filter is a single field filter. e.g. (a == 10). */
+  private static boolean isFieldFilter(Filter filter) {
+    return filter instanceof FieldFilter;
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/util/LogicUtils.java:43]: I would prefer if we also renamed the class.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/util/LogicUtils.java [hunk 11]
```diff
@@ -24,6 +30,265 @@
...
+    // Result: (A | B | C | D)
+    // Note that the `compositeSubfilter` has been eliminated, and its filters (B, C, D) have been
+    // added to the top-level "compositeFilter".
+    List<Filter> newSubfilters = new ArrayList<>();
+    for (Filter subfilter : updatedFilters) {
+      if (subfilter instanceof FieldFilter) {
+        newSubfilters.add(subfilter);
+      } else if (subfilter instanceof CompositeFilter) {
+        CompositeFilter compositeSubfilter = (CompositeFilter) subfilter;
+        if (compositeSubfilter.getOperator() == compositeFilter.getOperator()) {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/util/LogicUtils.java:143]: Prefer equals() even for enums.

**Review by schmidt-sebastian — APPROVED** [review]

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteIndexManager.java (hunk 9 above)
Comment (ehsannas) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteIndexManager.java]: Done

File: firebase-firestore/src/main/java/com/google/firebase/firestore/util/LogicUtils.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (ehsannas) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/util/LogicUtils.java:43]: you mean renaming the `FieldFilter` class to `SingleFieldFilter` ? (if so, it'll have to be done in another PR as it'd be a big change)

File: firebase-firestore/src/main/java/com/google/firebase/firestore/util/LogicUtils.java (hunk 11 above)
Comment (ehsannas) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/util/LogicUtils.java]: Done

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (ehsannas) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java]: Ah I see. Done.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/util/LogicUtils.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/util/LogicUtils.java:43]: Yes, and a follow up PR sounds fine.

**Pull request merged**